tools/nvmecmd.py

```diff
@@ -0,0 +1,237 @@
+  nvmecmd --ops 0                  # Observe read commands on all NVMe
+  nvmecmd --ops 1                  # Observe write commands on all NVMe
+  nvmecmd --ops 1 --disk nvme9n1   # Observe write commands on 9th NVMe node
+  nvmecmd --debug                  # Print eBPF program before observe
+  nvmecmd --trace                  # Print NVMe captured events
+  nvmecmd --interval 0.1           # Poll data ring buffer every 100 ms
+"""
+
+parser = argparse.ArgumentParser(
+    description="NVMe commands observer tool",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples,
+)
+parser.add_argument(
+    "-d",
+    "--disk",
+    type=str,
+    help="capture commands for this NVMe disk node only"
+)
+parser.add_argument(
+    "-o",
+    "--ops",
+    type=int,
+    help="capture this command operation only"
+)
+parser.add_argument("--debug", action="store_true", help="debug")
+parser.add_argument(
+    "--trace",
+    action="store_true",
+    help="trace NVMe captured commands"
+)
+parser.add_argument(
+    "--interval",
+    type=float,
+    help="polling interval"
+)
+
+args = parser.parse_args()
+
+level = logging.INFO
+if args.debug:
+    level = logging.DEBUG
+
+logging.basicConfig(level=level)
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/blk-mq.h>
+#include <linux/blk_types.h>
+#include <linux/nvme.h>
+
+struct nvme_ns {
+    struct list_head list;
+
+    struct nvme_ctrl *ctrl;
+    struct request_queue *queue;
+    struct gendisk *disk;
+#ifdef CONFIG_NVME_MULTIPATH
+    enum nvme_ana_state ana_state;
+    u32 ana_grpid;
+#endif
+    struct list_head siblings;
+    struct kref kref;
+    struct nvme_ns_head *head;
+
+    int lba_shift;
+    // [...]
+};
+
+struct data_t {
+    u32 pid;
+    char comm[TASK_COMM_LEN];
+    char disk[DISK_NAME_LEN];
+    u32 op;
+    u32 len;
+    u32 lba;
+    u32 algn;
+};
+
+BPF_HISTOGRAM(block_len, u32, 64);
+BPF_HISTOGRAM(algn, u32, 64);
+BPF_ARRAY(counts, u64, 1);
+BPF_RINGBUF_OUTPUT(events, 8);
+
+/* local strcmp function, max length 16 to protect instruction loops */
+#define CMPMAX	16
+
+static int local_strcmp(const char *cs, const char *ct)
+{
+    int len = 0;
+    unsigned char c1, c2;
+
+    while (len++ < CMPMAX) {
+        c1 = *cs++;
+        c2 = *ct++;
+        if (c1 != c2)
+            return c1 < c2 ? -1 : 1;
+        if (!c1)
+            break;
+    }
+    return 0;
+}
+"""
+
+bpf_text_disk_filter = ""
+if args.disk:
+    bpf_text_disk_filter = """
+        if (local_strcmp(req->q->disk->disk_name, "{disk}"))
+            return;
+    """.format(
+        disk=args.disk
+    )
+
+bpf_text_ops_filter = ""
+if args.ops:
+    bpf_text_ops_filter = """
+        if ((req->cmd_flags & 0xff) != {ops})
+            return;
+    """.format(
+        ops=args.ops
+    )
+
+bpf_text += """
+void kprobe__nvme_setup_cmd(struct pt_regs *ctx, struct nvme_ns *ns,
+                            struct request *req)
+{{
```

Comment:
Why '{{'?

Reply:
The second bracket is an escape bracket for the `{disk_filter}` and `{ops_filter}` variable replacement. This adds a disk and operation filter when specified by `--disk` and/or `--ops`.